### PR TITLE
[SYCL] Fix the type trait 'known_identity'

### PR DESCRIPTION
**Pull request by v-klochkov**
- Copy 'known_identity' and 'has_known_identity' from sycl::ONEAPI to sycl
- Fix the compilation errors reported on attempts to use vector types
  like sycl::int4 as a reduction variable.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

**Comment by v-klochkov** [comment]
This patch only fixes a correctness problem caused by using vec<int,N> types.
Potentially, known_identity< std::plus<>, int4 > could return a constexpr value, but that would require additional elaborate changes in vec class, and that is a lower priority task than the stability crash fixed in this pull request.

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp [hunk 1]
```diff
@@ -1632,16 +1611,18 @@ reduction(T *VarPtr, const T &Identity, BinaryOperation BOp) {
...
 reduction(T *VarPtr, BinaryOperation) {
   return detail::reduction_impl<T, BinaryOperation, 0, true,
                                 access::mode::read_write>(VarPtr);
 }
 
+} // namespace ONEAPI
+
 template <typename BinaryOperation, typename AccumulatorT>
-struct has_known_identity : detail::has_known_identity_impl<
+struct has_known_identity : ONEAPI::detail::has_known_identity_impl<
```
Comment (Pennycook) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:1625]: I don't know the answer to this, but I want to make sure that we're being consistent across different extensions and SYCL 2020 features.  Should this be available as `ONEAPI::has_known_identity` in SYCL 1.2.1 mode, and only available as `has_known_identity` in SYCL 2020 mode?

**Review by Pennycook — COMMENTED** [review]
I'm okay with this change as a temporary fix, but we need to revisit this issue.  It's not clear to me whether `has_known_identity` requiring `constexpr` is too restrictive, or whether types like `vec` should be made `constexpr`.

The SYCL 2020 specification doesn't seem to say whether an implementation is expected to be able to determine the identity value for vector types, and it probably comes down to whether or not a vector counts as "an arithmetic type" (see https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reduction).  This probably requires some clarification.

**Comment by rolandschulz** [comment]
> I'm okay with this change as a temporary fix, but we need to revisit this issue. It's not clear to me whether `has_known_identity` requiring `constexpr` is too restrictive, or whether types like `vec` should be made `constexpr`.

It might make sense to only support marray instead of vec and make sure that marray is constexpr.

**Comment by keryell** [comment]
Now SYCL 2020 is out we have more time to `constexpr`-ify the specification.

But I have the feeling that adding some `constexpr` today in an implementation should not change the compliance test and would make the implementation better, exposing more optimization opportunities.

I was about creating an issue about it but discovered I already created one... https://gitlab.khronos.org/sycl/Specification/-/issues/246 Ahem... :-/

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: > only available as has_known_identity in SYCL 2020 mode

John, I tried implementing what you suggested here: 9b355efa81966655860c15d8d88c0a0a393d3bcd
When you say "only in SYCL 2020 mode" what do you mean? I thought about using macro SYCL_LANGUAGE_VERSION automatically set with "-fsycl", but that now seems a bit too severe to me as in this case it is not possible to use has_known_identity without using -fsycl switch (and without explicit define of SYCL_LANGUAGE_VERSION macro).

Comment (Pennycook) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: I was assuming that we would still allow developers to request SYCL 1.2.1 -- if that isn't the case, then I'm less worried about conditionally enabling SYCL 2020 features.

That said, I think standard C++ headers typically guard features based on the standard that included them.  I don't feel strongly about this -- @rolandschulz, do you think it matters at all?

Comment (rolandschulz) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: If we don't (partially) support multiple SYCL versions, I don't see a reason to have features guarded.

**Review by Pennycook — APPROVED** [review]

**Comment by smaslov-intel** [comment]
> Move 'known_identity' and 'has_known_identity' from sycl::ONEAPI to sycl

Isn't this is a backward compatibility breaking change that we aren't allowed to do except in major releases?

**Comment by v-klochkov** [comment]
> > Move 'known_identity' and 'has_known_identity' from sycl::ONEAPI to sycl
> 
> Isn't this is a backward compatibility breaking change that we aren't allowed to do except in major releases?

The subsequence commit in this PR puts it back to sycl::ONEAPI. So, I'll replace the word 'move' to 'copy' in the description.

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp [hunk 2]
```diff
@@ -1659,7 +1642,30 @@ template <typename BinaryOperation, typename AccumulatorT>
 inline constexpr AccumulatorT known_identity_v =
     known_identity<BinaryOperation, AccumulatorT>::value;
 #endif
-
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:1662]: I think it is worth to mark the code that we want to remove when allowed to break API.

**Review by smaslov-intel — APPROVED** [review]
LGTM, just minor ask for a TODO comment to remember to remove the ONEAPI copy.

**Comment by v-klochkov** [comment]
> for a TODO comment to remember to remove the ONEAPI

ONEAPI::has_known_identity should live as long as ONEAPI::reduction() lives.

sycl::has_known_identity will migrate from the file CL/sycl/ONEAPI/reduction.hpp to CL/sycl/reduction.hpp as soon as SYCL-2020 reductions works start, and it will be tied to sycl::reduction.

**Pull request merged**